python/python/lance/arrow.py

```diff
@@ -11,14 +11,182 @@
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageURIType()
+
+    def __arrow_ext_class__(self):
+        return ImageURIArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageURIScalar
+
+
+class EncodedImageType(pa.ExtensionType):
+    def __init__(self):
+        pa.ExtensionType.__init__(self, pa.binary(), "lance.arrow.encoded_image")
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return EncodedImageType()
+
+    def __arrow_ext_class__(self):
+        return EncodedImageArray
+
+    def __arrow_ext_scalar_class__(self):
+        return EncodedImageScalar
+
+
+class ImageTensorType(pa.ExtensionType):
+    def __init__(self, arrow_type, shape):
+        # TODO: use pa.VariableShapeTensorArray once available?
+        pa.ExtensionType.__init__(
+            self, pa.fixed_shape_tensor(arrow_type, shape), "lance.arrow.image_tensor"
+        )
+
+    def __arrow_ext_serialize__(self):
+        return b""
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, self, storage_type, serialized):
+        return ImageTensorType()
+
+    def __arrow_ext_class__(self):
+        return ImageTensorArray
+
+    def __arrow_ext_scalar_class__(self):
+        return ImageTensorScalar
+
+
+class ImageURIArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.ImageURIArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, List[str]]):
+        if isinstance(uris, list):
+            return cls.from_storage(ImageURIType(), pa.array(uris, type=pa.string()))
+        else:
+            return cls.from_storage(ImageURIType(), uris)
+
+    def materialize(self):
+        # Assume first URI's filesystem is the same for all URIs
+        from pyarrow import fs
+
+        first_uri = self.storage[0].as_py()
+        filesystem, _ = fs.FileSystem.from_uri(first_uri)
+
+        # Read all URIs and return their content as lance.arrow.EncodedImageArray
+        images = []
+        for uri in self.storage:
+            with filesystem.open_input_stream(uri.as_py()) as f:
+                images.append(f.read())
+
+        return EncodedImageArray.from_binary_array(pa.array(images, type=pa.binary()))
+
+
+class EncodedImageArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.EncodedImageArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_binary_array(cls, array: pa.binary()):
+        return cls.from_storage(EncodedImageType(), array)
+
+    def materialize(self):
+        # TODO: try other decoders (pillow/torchvision/cv2/scipy/..) to decode images
+        # or rather just pick one and make it a requirement?
```

Comment:
Perhaps we can make `decoder` an argument to this method, so users can always pass in their own decoder?

Reply:
Passing a decoder like that would indeed be the cleanest option.
If we wanted to reduce dependencies we could vendor a rust image encoder/decoder (e.g. https://github.com/image-rs/image), but that would add quite some complexity.

Reply:
I think we should make these all optional dependencies, and we should instead just error if none of the auto-matically detected ones are installed with something like:

```python
raise ValueError("No image decoder installed. Please either install one of tensorflow, pillow, or X, or pass a decoder argument")
```

Reply:
Done. I've also used the same pattern for the encoding, but I'm no longer sure we really need encoding.